File: HeartForCharity/HeartForCharity.Services/Database/VolunteerApplication.cs
using HeartForCharity.Model.Enums;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace HeartForCharity.Services.Database
{
    public class VolunteerApplication
    {
        [Key]
        public int VolunteerApplicationId { get; set; }

        [Required]
        [ForeignKey(nameof(VolunteerJob))]
        public int VolunteerJobId { get; set; }

        [Required]
        [ForeignKey(nameof(UserProfile))]
        public int UserProfileId { get; set; }

        [MaxLength(4000)]
        public string? CoverLetter { get; set; }

        [MaxLength(500)]
        public string? ResumeUrl { get; set; }

        public ApplicationStatus Status { get; set; } = ApplicationStatus.Pending;

        [MaxLength(500)]
        public string? RejectionReason { get; set; }

        public bool IsCompleted { get; set; } = false;  // volonterski posao završen

        public int? ReviewedByUserId { get; set; }
        public DateTime? ReviewedAt { get; set; }

        public DateTime AppliedAt { get; set; } = DateTime.UtcNow;
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

        // Navigacijska svojstva
        public virtual VolunteerJob VolunteerJob { get; set; } = null!;
        public virtual UserProfile UserProfile { get; set; } = null!;
        [ForeignKey(nameof(ReviewedByUserId))]
        public virtual User? ReviewedByUser { get; set; }
        public virtual Review? Review { get; set; }
        public virtual ICollection<Notification> Notifications { get; set; } = new List<Notification>();
    }
}

File: HeartForCharity/UI/heartforcharity_mobile/lib/model/responses/volunteer_application.dart
class VolunteerApplication {
  final int volunteerApplicationId;
  final int volunteerJobId;
  final String? jobTitle;
  final int? userProfileId;
  final String? applicantName;
  final String? email;
  final String? phoneNumber;
  final DateTime? dateOfBirth;
  final String? address;
  final String? coverLetter;
  final String? resumeUrl;
  final String? status;
  final String? rejectionReason;
  final bool isCompleted;
  final bool hasReview;
  final String? reviewedByName;
  final DateTime? reviewedAt;
  final DateTime? appliedAt;
  final DateTime? updatedAt;

  VolunteerApplication({
    this.volunteerApplicationId = 0,
    this.volunteerJobId = 0,
    this.jobTitle,
    this.userProfileId,
    this.applicantName,
    this.email,
    this.phoneNumber,
    this.dateOfBirth,
    this.address,
    this.coverLetter,
    this.resumeUrl,
    this.status,
    this.rejectionReason,
    this.isCompleted = false,
    this.hasReview = false,
    this.reviewedByName,
    this.reviewedAt,
    this.appliedAt,
    this.updatedAt,
  });

  factory VolunteerApplication.fromJson(Map<String, dynamic> json) =>
      VolunteerApplication(
        volunteerApplicationId: json['volunteerApplicationId'] ?? 0,
        volunteerJobId: json['volunteerJobId'] ?? 0,
        jobTitle: json['jobTitle'],
        userProfileId: json['userProfileId'],
        applicantName: json['applicantName'],
        email: json['email'],
        phoneNumber: json['phoneNumber'],
        dateOfBirth: json['dateOfBirth'] != null ? DateTime.parse(json['dateOfBirth']) : null,
        address: json['address'],
        coverLetter: json['coverLetter'],
        resumeUrl: json['resumeUrl'],
        status: json['status'],
        rejectionReason: json['rejectionReason'],
        isCompleted: json['isCompleted'] ?? false,
        hasReview: json['hasReview'] ?? false,
        reviewedByName: json['reviewedByName'],
        reviewedAt: json['reviewedAt'] != null ? DateTime.parse(json['reviewedAt']) : null,
        appliedAt: json['appliedAt'] != null ? DateTime.parse(json['appliedAt']) : null,
        updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      );
}
